README.md
## Go Share

```ASCII
                        __
   ____ _____     _____/ /_  ____ _________
  / __ `/ __ \   / ___/ __ \/ __ `/ ___/ _ \
 / /_/ / /_/ /  (__  ) / / / /_/ / /  /  __/
 \__, /\____/  /____/_/ /_/\__,_/_/   \___/
/____/

```

[Tasks in Queue at Trello Board](https://trello.com/b/ZjDMRGQN/goshare)

#### Go Share any data among the nodes. Over HTTP or ZeroMQ.

* GOShare eases up communication over HTTP GET param based interaction.
* ZeroMQ REQ/REP based synchronous communication model.

it's "go get"-able

``` go get "github.com/abhishekkr/goshare" ```

***

#### Make Distributable Binary

```bash
./go-tasks.sh bin
```

This will create two distributable binaries *./bin/goshare_service* & *./bin/goshare_daemon*. Here *./bin/goshare_service* works as shown in README's 'Tryout' section.

Whereas *./bin/goshare_daemon* can be used as a system service daemon, with following command line flags (along with flags mentioned in 'Tryout' section for ports and db-path)
>
> * start: ``` ./bin/goshare_daemon -daemon=start ```
> * stop: ``` ./bin/goshare_daemon -daemon=stop ```
> * status: ``` ./bin/goshare_daemon -daemon=status ```
>
> *this dumps daemon's current status to /tmp/goshare_daemon.status and pid to /tmp/goshare_daemon.pid*
> *the status and pid file path can be changed with flags '-daemon-log=<path>' & '-daemon-pid=<path>' respectively*
>


#### Tryout:

```Shell
 go run zxtra/goshare_daemon.go -dbpath=/tmp/GOTSDB
```

By default it runs HTTP daemon at port 9999 and ZeroMQ daemon at 9797/9898,
make it run on another port using following required flags

```Shell
 go run zxtra/goshare_daemon.go -dbpath=/tmp/GOTSDB -port=8080 -req-port=8000 -rep-port=8001
```

```ASCII
  Dummy Clients Using It

  * go run zxtra/gohttp_client.go

  * go run zxtra/go0mq_client.go


  for custom Port: 8080 for HTTP; Port: 8000/8001 for ZeroMQ

  * go run zxtra/gohttp_client.go -port=8080

  * go run zxtra/go0mq_client.go -req-port=8000 -rep-port=8001
```

>
> To utilize it "zxtra/gohttp_client.go" and "zxtra/go0mq_client.go" can be referred on how to utilize capabilities of GoShare.
>

***

#### Structure:

> "goshare"'s methods to adapt these in your code:
>
> * GoShare() : it runs HTTP and ZeroMQ daemon in parallel goroutines
> > has optional flags customization of:
> > * dbpath: path for LevelDB (default: /tmp/GO.DB)
> > * port: port to bind HTTP daemon (default: 9999)
> > * req-port, rep-port: ports to bind ZeroMQ REQ/REP daemon (default: 9797, 9898)
>
> * GoShareHTTP(&lt;levigo DB handle&gt;, &lt;http port as int&gt;) : it runs HTTP daemon
>
> * GoShareZMQ(&lt;levigo DB handle&gt;, &lt;req-port as int&gt;, &lt;rep-port as int&gt;) : it runs ZMQ daemon
>

***

Now visit the the link asked by it and get the help page.

##### Dependency
* [go lang](http://golang.org/doc/install) (obviously, the heart and soul of the app)
* [leveldb](http://en.wikipedia.org/wiki/LevelDB) (we are using for datastore, it's awesome)
* [levigo](https://github.com/jmhodges/levigo/blob/master/README.md) (the go library utilized to access leveldb)
* [zeroMQ](http://zeromq.org/) (the supercharged Sockets giving REQuest/REPly power)
* [gozmq](https://github.com/alecthomas/gozmq) GoLang ZeroMQ Bindings used here
* [levigoNS](https://github.com/abhishekkr/levigoNS) NameSpace KeyVal capabilities around leveldb via levigo
* [levigoTSDS](https://github.com/abhishekkr/levigoTSDS) TimeSeries KeyVal capabilties around leveldb via levigoNS
* [gol](https://github.com/abhishekkr/gol) Set of common utility functionalities

[![baby-gopher](https://raw2.github.com/drnic/babygopher-site/gh-pages/images/babygopher-badge.png)](http://www.babygopher.org)


CHANGELOG.md
## Changelog

### v0.8.5

* [db-tasks] moved to use golkeyval{,NS,TSDS} for switch-able db back-end
* golint-ified goshare base code


### v0.8.4

* [read,delete] enable {def,ns,tsds}-{,csv,json} to get multiple keys, this also define multi-val type for "read" output

#### v0.8.3

* ParentNS feature provided, provide Parent NameSpace for all keys (in any key-type) to be used while Push/Read/Delete
* dbtasks moved to use Packet{}, abstract-ing out how data is received on multiple communication channels

#### v0.8.2

* [push] encoding for ns/tsds may contain a root parent-ns config to pre-pend for all ASCII multi-val

---

#### v0.8.1

* added json support
* fixed multi-val support for non-tsds key-types

#### v0.8.0

* huge refactor and design fixes

#### *Good To Have*

>
> * fetch all one-step {,all with val::} children in a particular key namespace, don't fetch all with-val:: children in a particular key namespace, 'cuz that can be done by doing ns and looking at all keys
>
> * (logging) "debug:default" log triggers only at failure status; "verbose" with only DBRest call logged again
>
> * [read] features for tsds like: {latest,this}_{year,month,week,day,hour,hour,min,dot}
>
> * [read] for tsds: mean, median, max, min, from_X_to_Y, radius_X_of_Y, more_than_X, less_than_X, matching_to_X, not_matching_to_X
>

#### *Design Decisions*

>
> * if task-type tokens need to grow to more than 3, move to msgpack for entire packet sent to dbtasks
>


flag_handler.go
package goshare

import (
	"flag"
	"fmt"

	"github.com/abhishekkr/gol/golconfig"
)

/*
Config is a hashmap used here to carry around param=paramValue for GoShare.
*/
type Config map[string]string

// flags
var (
	flagConfig     = flag.String("config", "", "the path to overriding config file")
	flagDBEngine   = flag.String("DBEngine", "leveldb", "the type of KeyVal DB backend to be used")
	flagDBPath     = flag.String("DBPath", "/tmp/GO.DB", "the path to DB")
	flagServerUri  = flag.String("server-uri", "0.0.0.0", "what Port to Run HTTP Server at")
	flagHTTPPort   = flag.String("http-port", "9999", "what Port to Run HTTP Server at")
	flagRepPorts   = flag.String("rep-ports", "9898,9797", "what PORT to run ZMQ REP at")
	flagCPUProfile = flag.String("cpuprofile", "", "write cpu profile to file")
)

/* assignIfEmpty assigns val to *key only if it's empty */
func assignIfEmpty(mapper Config, key string, val string) {
	if mapper[key] == "" {
		mapper[key] = val
	}
}

/*
ConfigFromFlags configs from values provided to flags.
*/
func ConfigFromFlags() Config {
	flag.Parse()

	var config Config
	config = make(Config)
	if *flagConfig != "" {
		configFile := golconfig.GetConfig("json")
		configFile.ConfigFromFile(*flagConfig, &config)
	}

	assignIfEmpty(config, "DBEngine", *flagDBEngine)
	assignIfEmpty(config, "DBPath", *flagDBPath)
	assignIfEmpty(config, "server-uri", *flagServerUri)
	assignIfEmpty(config, "http-port", *flagHTTPPort)
	assignIfEmpty(config, "rep-ports", *flagRepPorts)
	assignIfEmpty(config, "cpuprofile", *flagCPUProfile)

	fmt.Println("Starting for:", config)
	return config
}
